src/vdom/builder.rs

```diff
@@ -1,128 +1,218 @@
-use web_sys::Element;
+use std::marker::PhantomData;
 
-use super::{alias::*, App};
+use crate::{
+    dom_types::View,
+    orders::OrdersContainer,
+    routing,
+    vdom::{alias::*, App, AppRunCfg},
+};
 
-use crate::{dom_types::View, orders::OrdersContainer, routing, util};
+pub mod init;
+pub use init::{Fn as InitFn, Init, Into as IntoInit};
+pub mod before_mount;
+pub use before_mount::{BeforeMount, Into as IntoBeforeMount, MountPoint, MountType};
+pub mod after_mount;
+pub use after_mount::{AfterMount, Into as IntoAfterMount, UrlHandling};
 
-pub type InitFn<Ms, Mdl, ElC, GMs> =
-    Box<dyn FnOnce(routing::Url, &mut OrdersContainer<Ms, Mdl, ElC, GMs>) -> Init<Mdl>>;
-
-pub trait MountPoint {
-    fn element(self) -> Element;
+pub trait InitAPI<Ms: 'static, Mdl, ElC: View<Ms>, GMs> {
+    type Builder;
+    fn build(builder: Self::Builder) -> App<Ms, Mdl, ElC, GMs>;
 }
 
-impl MountPoint for &str {
-    fn element(self) -> Element {
-        util::document().get_element_by_id(self).unwrap_or_else(|| {
-            panic!(
-                "Can't find element with id={:?} - app cannot be mounted!\n\
-                 (Id defaults to \"app\", or can be set with the .mount() method)",
-                self
-            )
-        })
-    }
+pub struct MountPointInitInitAPI<MP, II> {
+    mp: PhantomData<MP>,
+    ii: PhantomData<II>,
```

Comment:
Can use more expressive names? I have no idea what `mp` and `ii` are without context.

Reply:
Hrm, yeah, true. I might end up moving the 4 fields into their respective `InitAPI` structs, so this might change a bit today.